Sources/XcodesKit/Environment.swift

```diff
@@ -53,6 +53,75 @@ public struct Shell {
     public func xcodeSelectSwitch(password: String?, path: String) -> Promise<ProcessOutput> {
         xcodeSelectSwitch(password, path)
     }
+    
+    public var downloadWithAria2: (Path, URL, Path, [HTTPCookie]) -> (Progress, Promise<Void>) = { aria2Path, url, destination, cookies in
+        let process = Process()
+        process.executableURL = aria2Path.url
+        process.arguments = [
+            "--header=Cookie: \(cookies.map { "\($0.name)=\($0.value)" }.joined(separator: "; "))",     
+            "--max-connection-per-server=16",
+            "--split=16",
+            "--summary-interval=1",
+            "--stop-with-process=\(ProcessInfo.processInfo.processIdentifier)",
+            "--dir=\(destination.parent.string)",
+            "--out=\(destination.basename())",
+            url.absoluteString,
+        ]
+        let stdOutPipe = Pipe()
+        process.standardOutput = stdOutPipe
+        let stdErrPipe = Pipe()
+        process.standardError = stdErrPipe
+        
+        var progress = Progress(totalUnitCount: 100)
+
+        let token = NotificationCenter.default.addObserver(
+            forName: .NSFileHandleDataAvailable, 
+            object: nil, 
+            queue: OperationQueue.main
+        ) { note in
+            let handle = note.object as! FileHandle
+            guard handle === stdOutPipe.fileHandleForReading || handle === stdErrPipe.fileHandleForReading else { return }
+            defer { handle.waitForDataInBackgroundAndNotify() }
+            let string = String(decoding: handle.availableData, as: UTF8.self)
+            let regex = try! NSRegularExpression(pattern: #"((?<percent>\d+)%\))"#)
+            let range = NSRange(location: 0, length: string.utf16.count)
+            guard
+                let match = regex.firstMatch(in: string, options: [], range: range),
+                let matchRange = Range(match.range(withName: "percent"), in: string),
+                let percentCompleted = Int64(string[matchRange])
+            else { return }
+            progress.completedUnitCount = percentCompleted
+        }
+
+        stdOutPipe.fileHandleForReading.waitForDataInBackgroundAndNotify()
+        stdErrPipe.fileHandleForReading.waitForDataInBackgroundAndNotify()
+        
+        do {
+            try process.run()
+        } catch {
+            return (progress, Promise(error: error))
+        }
+
+        let promise = Promise<Void> { seal in
+            // Capture token 
+            _ = token
```

Comment:
I don't understand what this is doing.

Reply:
Mmm, it's wrong anyways, thanks for making me check! 😄 The intent was to capture `token` to keep the observation alive while the promise was, but block-based observation still requires removing the observer (unlike selector-based observation.) I've changed this to remove the token after the promise is resolved.